{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3e745105",
   "metadata": {},
   "source": [
    "# Counting reads from bam files with ReadCounter\n",
    "\n",
    "Sincei provides the [scCountReads](../tools/scCountReads.rst) command line tool to produce count matrices stored in\n",
    "``AnnData`` format. This tool is based on the [sincei.ReadCounter](../modules/ReadCounter.rst) module\n",
    "which can be used directly in python.\n",
    "\n",
    "This tutorial demonstrates how to use [sincei.ReadCounter](../modules/ReadCounter.rst) to produce\n",
    "read count matrices from bam files. We will use the example data provided in our\n",
    "[sc-sortChIC analysis tutorial](https://sincei.readthedocs.io/en/latest/content/tutorials/sincei_tutorial_sortChIC.html)\n",
    "and our [10x multiome scRNA-seq analysis tutorial](https://sincei.readthedocs.io/en/latest/content/tutorials/sincei_tutorial_10xRNA.html).\n",
    "\n",
    "Both datasets can be downloaded from figshare: [sc-sortChIC](https://figshare.com/articles/dataset/sortChIC_testdata_package/23544774) and [10x sc-RNAseq](https://figshare.com/articles/dataset/10x_multiome_test_data_package/29424470?file=60078353)\n",
    "\n",
    "You can download this tutorial as a jupyter notebook by clicking the icon on the top right bar.\n",
    "\n",
    "The ``ReadCounter`` module contains a main class, [CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin)\n",
    "that takes the paths to our bamfiles as input, and returns a _feature x cell_ count matrix in ``numpy.ndarray`` format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6bf4e854",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "import anndata as ad\n",
    "\n",
    "from sincei.ReadCounter import CountReadsPerBin\n",
    "\n",
    "threads = 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d7ca5cd",
   "metadata": {},
   "source": [
    "## 1. Counting reads in genomic bins\n",
    "\n",
    "First, we will produce a count matrix for a subset of the _single-cell sortChIC H3K27me3_ data from\n",
    "[Zeller, Yueng et. al. (2023)](https://www.nature.com/articles/s41588-022-01260-3). We aggregate the signal\n",
    "in 10kb bins.\n",
    "\n",
    "Our data is located in the ``sortchic_testdata`` directory. We can store the paths to our bamfiles and\n",
    "to the file containing our cell barcodes using the following commands. Note that the cell barcodes must\n",
    "be stored in a list of strings, where each string corresponds to one barcode.\n",
    "\n",
    "Additionally, we need a blacklist of problematic regions in the genome to filter out sequencing artifacts.\n",
    "We provide a blacklist for the mm10 mouse genome in our example data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b4d47ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = 'sortchic_testdata'\n",
    "bam_files = [bam_file.path for bam_file in os.scandir(data_dir) if bam_file.path.endswith('.bam')]\n",
    "\n",
    "barcodes_path = os.path.join(data_dir, 'sortChIC_barcodes.txt')\n",
    "barcodes = [line.strip() for line in open(barcodes_path)]\n",
    "\n",
    "blacklist = os.path.join(data_dir, 'mm10_blacklist.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e431693e",
   "metadata": {},
   "source": [
    "[CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin) takes the following arguments:\n",
    "- The list of paths to the bamfiles\n",
    "- Optionally, a region to limit the counting to in the format ``chrom:start:end``\n",
    "- The length of the bins to aggregate the signal into\n",
    "- The step size between bins. This is the distance between the start of two consecutive bins.\n",
    "Set equal to ``binLength`` to have non-overlapping bins covering the whole region.\n",
    "- The list of barcodes to count cells for, where each barcode corresponds to a cell in each sample.\n",
    "- The BAM tag where the cell barcodes are stored\n",
    "- The path to the blacklist BED/GTF file\n",
    "- The number of cores to use in the counting\n",
    "\n",
    "We instantiate [CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin) with\n",
    "all the necessary parameters, and perform the counting by calling ``CountReadsPerBin.run()``.\n",
    "This returns our _bin x cell_ count matrix and numpy array of bin IDs.\n",
    "\n",
    "Here we count the reads in 50kb bins across chromosome 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c439b6f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellTag = 'BC'\n",
    "region = 'chr1'\n",
    "bin_length = 50000\n",
    "\n",
    "counter_bins = CountReadsPerBin(\n",
    "    bam_files,\n",
    "    region=region,\n",
    "    binLength=bin_length,\n",
    "    stepSize=bin_length,\n",
    "    barcodes=barcodes,\n",
    "    cellTag=cellTag,\n",
    "    blackListFileName=blacklist,\n",
    "    numberOfProcessors=threads,\n",
    "    )\n",
    "bin_counts, bin_IDs = counter_bins.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d0876ba6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Counted 1536 barcodes and 3923 in chr1.\n",
      "Total counts: 14754957.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"\"\"\n",
    "Counted {bin_counts.shape[1]} barcodes and {bin_counts.shape[0]} in chr1.\n",
    "Total counts: {bin_counts.sum()}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e70ff916",
   "metadata": {},
   "source": [
    "Be mindful that each column output matrix contains the data for one barcode in each bamfile. Some\n",
    "barcodes may not be present, or empty of counts, so filtering the count matrix is required to keep\n",
    "only barcodes corresponding to real cells. We provide a CLI tool for this purpose,\n",
    "[scCountQC](../tools/scCountQC.rst), and an example of its usage in\n",
    "[this tutorial](sincei_tutorial_sortChIC.rst).\n",
    "\n",
    "We can store our count matrix in an AnnData object for later processing and analysis.\n",
    "\n",
    "Note that ``AnnData`` stores data in _cell x feature_ format, so we must transpose the data matrix\n",
    "when assigning it to ``adata.X``. Additionally, we can use ``bam_files``, ``barcodes``, and\n",
    "``bin_IDs`` to add columns to the observation and variable dataframes of our ``AnnData.``"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8d7d9680",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Observation names and columns\n",
    "sample_names = [Path(bam_file).stem for bam_file in bam_files]\n",
    "samples = [sample for sample in sample_names for _ in range(len(barcodes))]\n",
    "cell_IDs = [f\"{sample}::{barcode}\" for sample in sample_names for barcode in barcodes]\n",
    "\n",
    "# Variable names and columns\n",
    "var_names = [var.split('::')[0] for var in bin_IDs]\n",
    "chroms = [var.split('_')[0] for var in var_names]\n",
    "starts = [int(var.split('_')[1]) for var in var_names]\n",
    "ends = [int(var.split('_')[2]) for var in var_names]\n",
    "\n",
    "# Create anndata object\n",
    "sortChIC_adata = ad.AnnData(\n",
    "    X=bin_counts.T,\n",
    "    obs={'sample': samples, 'barcode': barcodes * len(sample_names)},\n",
    "    var={'chrom': chroms, 'start': starts, 'end': ends},\n",
    "    )\n",
    "sortChIC_adata.obs_names = cell_IDs\n",
    "sortChIC_adata.var_names = bin_IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d367163e",
   "metadata": {},
   "source": [
    "We have given a unique ID to each of the cells in our dataset by concatenating the sample they come\n",
    "from (with `::` as a separator), and their individual cell barcode, as we can see below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c0e7f0ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACACTA</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACACTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACATAG</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACATAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACGAGA</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACGAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACTATC</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACTATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACTGAT</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACTGAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    sample   barcode\n",
       "sortChIC-BM-SL2-k4me1-1::ACACACTA  sortChIC-BM-SL2-k4me1-1  ACACACTA\n",
       "sortChIC-BM-SL2-k4me1-1::ACACATAG  sortChIC-BM-SL2-k4me1-1  ACACATAG\n",
       "sortChIC-BM-SL2-k4me1-1::ACACGAGA  sortChIC-BM-SL2-k4me1-1  ACACGAGA\n",
       "sortChIC-BM-SL2-k4me1-1::ACACTATC  sortChIC-BM-SL2-k4me1-1  ACACTATC\n",
       "sortChIC-BM-SL2-k4me1-1::ACACTGAT  sortChIC-BM-SL2-k4me1-1  ACACTGAT"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sortChIC_adata.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0531c76",
   "metadata": {},
   "source": [
    "The ID of each region, corresponds to their position in the chromosome concatenated with the name of\n",
    "each region. Since we counted in ``bins``, the regions lack a name, so it shows as _None_. In the\n",
    "section below, we will count reads from regions in  _BED/GTF_ files, where this will be relevant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3222d399",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_182050000_182100000::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182050000</td>\n",
       "      <td>182100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_182100000_182150000::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182100000</td>\n",
       "      <td>182150000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_182150000_182200000::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182150000</td>\n",
       "      <td>182200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_182200000_182250000::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182200000</td>\n",
       "      <td>182250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_182250000_182300000::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182250000</td>\n",
       "      <td>182300000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               chrom      start        end\n",
       "chr1_182050000_182100000::None  chr1  182050000  182100000\n",
       "chr1_182100000_182150000::None  chr1  182100000  182150000\n",
       "chr1_182150000_182200000::None  chr1  182150000  182200000\n",
       "chr1_182200000_182250000::None  chr1  182200000  182250000\n",
       "chr1_182250000_182300000::None  chr1  182250000  182300000"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sortChIC_adata.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e30d682f",
   "metadata": {},
   "source": [
    "## 2. Counting reads in genomic features\n",
    "\n",
    "In addition to aggregating signal in bins, we can count the read counts for a known feature set. We\n",
    "provide a BED/GTF file containing our regions of interest to\n",
    "[CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin)\n",
    "instead of ``binLength`` and ``stepSize``.\n",
    "\n",
    "We will produce a count matrix of scATAC-seq signal in chromatin accessibility peaks for a region of\n",
    "chromosome 1, and a scRNA-seq count matrix for the genes in the same region. The data comes from a\n",
    "subset of cells of the 10x multiome dataset published in\n",
    "[Persad et. al.Â (2023)](https://www.nature.com/articles/s41587-023-01716-9)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "17a33e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '10x_multiome_testdata'\n",
    "\n",
    "barcodes_path = os.path.join(data_dir, '10x_barcodes_tutorial.txt')\n",
    "barcodes = [line.strip() for line in open(barcodes_path)]\n",
    "\n",
    "blacklist = os.path.join(data_dir, 'hg38_blacklist.v2.bed')\n",
    "\n",
    "region = 'chr1:47000000:47300000'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e869018",
   "metadata": {},
   "source": [
    "### scATAC-seq counts\n",
    "The file ``atacPeaks_replicateMerged.bed``contains chromatin accessibility peaks. Using it as input\n",
    "for [CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin) overrides the\n",
    "bin counting parameters. The rows in the resulting matrix each correspond to one region in the file.\n",
    "\n",
    "We specify that counting should be performed only in our regions of interest, so only peaks within\n",
    "the region will be included."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9ac54fa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_files = [bam_file.path for bam_file in os.scandir(data_dir)\n",
    "             if ('atac' in bam_file.path) and bam_file.path.endswith('.bam')]\n",
    "\n",
    "peaks_bed = os.path.join(data_dir, 'atacPeaks_replicateMerged.bed')\n",
    "\n",
    "cellTag = 'CB'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4fb12c62",
   "metadata": {},
   "outputs": [],
   "source": [
    "counter_atac = CountReadsPerBin(\n",
    "    bam_files,\n",
    "    region = region,\n",
    "    bedFile = peaks_bed,\n",
    "    blackListFileName=blacklist,\n",
    "    barcodes = barcodes,\n",
    "    cellTag = cellTag,\n",
    "    numberOfProcessors=threads,\n",
    "    )\n",
    "atac_peak_counts, atac_peaks = counter_atac.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a0c624c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Counted 4256 barcodes and 45 peaks in chr1:47000000:47300000.\n",
      "Total counts: 32980.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"\"\"\n",
    "Counted {atac_peak_counts.shape[1]} barcodes and {atac_peak_counts.shape[0]} peaks in {region}.\n",
    "Total counts: {atac_peak_counts.sum()}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7b16ede",
   "metadata": {},
   "source": [
    "### scRNA-seq\n",
    "As with our with counting our scATAC-seq reads, we provide a file with regions for counting, in this\n",
    "case a GTF file containing all the genes in the HG38 human genome assembly. This file can be\n",
    "downloaded [here](http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.refGene.gtf.g).\n",
    "\n",
    "The file ``atacPeaks_replicateMerged.bed``contains chromatin accessibility peaks. Using it as input\n",
    "for [CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin) overrides the\n",
    "bin counting parameters. The rows in the resulting matrix each correspond to one region in the file.\n",
    "\n",
    "We specify that counting should be performed only in our regions of interest, so only peaks within\n",
    "the region will be included."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "52809d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_files = [bam_file.path for bam_file in os.scandir(data_dir)\n",
    "             if ('gex' in bam_file.path) and bam_file.path.endswith('.bam')]\n",
    "\n",
    "genes_gtf = os.path.join(data_dir, 'hg38.gtf')\n",
    "\n",
    "cellTag = 'CB'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fe7155e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "counter_genes = CountReadsPerBin(\n",
    "    bam_files,\n",
    "    region = region,\n",
    "    bedFile = genes_gtf,\n",
    "    blackListFileName=blacklist,\n",
    "    barcodes = barcodes,\n",
    "    cellTag = cellTag,\n",
    "    numberOfProcessors=threads,\n",
    "    )\n",
    "rna_gene_counts, genes = counter_genes.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "327e09f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Counted 4256 barcodes and 19 genes in chr1:47000000:47300000.\n",
      "Total counts: 9890.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"\"\"\n",
    "Counted {rna_gene_counts.shape[1]} barcodes and {rna_gene_counts.shape[0]} genes in {region}.\n",
    "Total counts: {rna_gene_counts.sum()}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f5df435",
   "metadata": {},
   "source": [
    "As mentioned in the section before, the names of our regions consist of their genomic position,\n",
    "followed by the region name specified in the BED/GTF file. We can see this below, the ID of each\n",
    "region is followed by its corresponding transcript ID. These can easily be converted into gene names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "77c28fd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr1_47004185_47050751::NM_001320290'\n",
      " 'chr1_47023668_47050751::NM_178033' 'chr1_47067230_47118318::NM_178134'\n",
      " 'chr1_47137424_47147481::NM_001308102'\n",
      " 'chr1_47137440_47149727::NM_001010969'\n",
      " 'chr1_47179249_47180339::NR_047498' 'chr1_47183581_47190036::NM_005764'\n",
      " 'chr1_47216289_47231715::NM_001287347'\n",
      " 'chr1_47216289_47231889::NM_001290404'\n",
      " 'chr1_47216289_47231889::NM_001290405'\n",
      " 'chr1_47216289_47232335::NM_001290403'\n",
      " 'chr1_47216289_47232335::NM_003189'\n",
      " 'chr1_47216290_47232225::NM_001290406'\n",
      " 'chr1_47250138_47314147::NM_001048166'\n",
      " 'chr1_47250138_47314147::NM_001282936'\n",
      " 'chr1_47250138_47314147::NM_001282937'\n",
      " 'chr1_47250138_47314147::NM_001282938'\n",
      " 'chr1_47250138_47314147::NM_001282939'\n",
      " 'chr1_47250138_47314147::NM_003035']\n"
     ]
    }
   ],
   "source": [
    "print(genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da7e8292",
   "metadata": {},
   "source": [
    "#### (Optional) Convert transcript IDs to gene IDs\n",
    "\n",
    "The following python function that reads a GTF file into a pandas dataframe. We will read\n",
    "`hg38.gtf` and use the resulting dataframe to convert our transcript IDs to gene IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "bc6fd207",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "def parse_GTF(gtf_path):\n",
    "\n",
    "    def _parse_attrs(attr):\n",
    "        d = {}\n",
    "        for m in re.finditer(r'\\s*([^;\\s]+)\\s+\"([^\"]+)\"', str(attr)):\n",
    "            d[m.group(1)] = m.group(2)\n",
    "        return d\n",
    "    \n",
    "    gtf_df = pd.read_csv(\n",
    "        gtf_path, sep='\\t', comment='#', header=None,\n",
    "        names=['seqname', 'source', 'feature', 'start', 'end',\n",
    "               'score', 'strand', 'frame', 'attribute'])\n",
    "    \n",
    "    attributes = gtf_df['attribute'].apply(_parse_attrs)\n",
    "    attributes_df = pd.DataFrame(list(attributes)).fillna('')\n",
    "\n",
    "    gtf_df = pd.concat([gtf_df.drop(columns=['attribute']), attributes_df], axis=1)\n",
    "\n",
    "    return gtf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a5740365",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>refGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>refGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>NR_046018.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>refGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>refGene</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>3</td>\n",
       "      <td>NR_046018.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>refGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname   source     feature  start    end score strand frame  gene_id  \\\n",
       "0    chr1  refGene  transcript  11874  14409     .      +     .  DDX11L1   \n",
       "1    chr1  refGene        exon  11874  12227     .      +     .  DDX11L1   \n",
       "2    chr1  refGene        exon  12613  12721     .      +     .  DDX11L1   \n",
       "3    chr1  refGene        exon  13221  14409     .      +     .  DDX11L1   \n",
       "4    chr1  refGene  transcript  14362  29370     .      -     .   WASH7P   \n",
       "\n",
       "  transcript_id gene_name exon_number      exon_id  \n",
       "0     NR_046018   DDX11L1                           \n",
       "1     NR_046018   DDX11L1           1  NR_046018.1  \n",
       "2     NR_046018   DDX11L1           2  NR_046018.2  \n",
       "3     NR_046018   DDX11L1           3  NR_046018.3  \n",
       "4     NR_024540    WASH7P                           "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df = parse_GTF(genes_gtf)\n",
    "gtf_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d55d0f0",
   "metadata": {},
   "source": [
    "The GTF dataframe contains the mapping between transcript IDs and gene IDs. By merging the list of\n",
    "our transcript IDs with the GTF dataframe, we can add make obtain the list of corresponding gene IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "8fc74655",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001320290</td>\n",
       "      <td>CYP4X1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_178033</td>\n",
       "      <td>CYP4X1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_178134</td>\n",
       "      <td>CYP4Z1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001308102</td>\n",
       "      <td>CYP4A22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001010969</td>\n",
       "      <td>CYP4A22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NR_047498</td>\n",
       "      <td>LINC00853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_005764</td>\n",
       "      <td>PDZK1IP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NM_001287347</td>\n",
       "      <td>TAL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NM_001290404</td>\n",
       "      <td>TAL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NM_001290405</td>\n",
       "      <td>TAL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NM_001290403</td>\n",
       "      <td>TAL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_003189</td>\n",
       "      <td>TAL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NM_001290406</td>\n",
       "      <td>TAL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_001048166</td>\n",
       "      <td>STIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NM_001282936</td>\n",
       "      <td>STIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NM_001282937</td>\n",
       "      <td>STIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_001282938</td>\n",
       "      <td>STIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NM_001282939</td>\n",
       "      <td>STIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NM_003035</td>\n",
       "      <td>STIL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   transcript_id  gene_name\n",
       "0   NM_001320290     CYP4X1\n",
       "1      NM_178033     CYP4X1\n",
       "2      NM_178134     CYP4Z1\n",
       "3   NM_001308102    CYP4A22\n",
       "4   NM_001010969    CYP4A22\n",
       "5      NR_047498  LINC00853\n",
       "6      NM_005764   PDZK1IP1\n",
       "7   NM_001287347       TAL1\n",
       "8   NM_001290404       TAL1\n",
       "9   NM_001290405       TAL1\n",
       "10  NM_001290403       TAL1\n",
       "11     NM_003189       TAL1\n",
       "12  NM_001290406       TAL1\n",
       "13  NM_001048166       STIL\n",
       "14  NM_001282936       STIL\n",
       "15  NM_001282937       STIL\n",
       "16  NM_001282938       STIL\n",
       "17  NM_001282939       STIL\n",
       "18     NM_003035       STIL"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_to_gene = pd.DataFrame({'transcript_id': [genes.split('::')[1] for genes in genes],})\n",
    "\n",
    "transcript_to_gene = transcript_to_gene.merge(\n",
    "    gtf_df[['transcript_id', 'gene_name']].drop_duplicates(),\n",
    "    on='transcript_id', how='left',\n",
    ")\n",
    "transcript_to_gene"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
