{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3e745105",
   "metadata": {},
   "source": [
    "# Counting reads from bam files with ReadCounter\n",
    "\n",
    "Sincei provides the [scCountReads](../tools/scCountReads.rst) command line tool to produce count matrices stored in\n",
    "``AnnData`` format. This tool is based on the [sincei.ReadCounter](../modules/ReadCounter.rst) module\n",
    "which can be used directly in python.\n",
    "\n",
    "This tutorial demonstrates how to use [sincei.ReadCounter](../modules/ReadCounter.rst) to produce\n",
    "read count matrices from bam files. We will use the example data provided in our\n",
    "[sc-sortChIC analysis tutorial](https://sincei.readthedocs.io/en/latest/content/tutorials/sincei_tutorial_sortChIC.html)\n",
    "and our [10x multiome scRNA-seq analysis tutorial](https://sincei.readthedocs.io/en/latest/content/tutorials/sincei_tutorial_10xRNA.html).\n",
    "\n",
    "Both datasets can be downloaded from figshare: [sc-sortChIC](https://figshare.com/articles/dataset/sortChIC_testdata_package/23544774) and [10x sc-RNAseq](https://figshare.com/articles/dataset/10x_multiome_test_data_package/29424470?file=60078353)\n",
    "\n",
    "You can download this tutorial as a jupyter notebook by clicking the icon on the top right bar.\n",
    "\n",
    "The ``ReadCounter`` module contains a main class, [CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin)\n",
    "that takes the paths to our bamfiles as input, and returns a _feature x cell_ count matrix in ``numpy.ndarray`` format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6bf4e854",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "import anndata as ad\n",
    "\n",
    "from sincei.ReadCounter import CountReadsPerBin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d7ca5cd",
   "metadata": {},
   "source": [
    "## 1. Counting reads in genomic bins\n",
    "\n",
    "First, we will produce a count matrix for a subset of the _single-cell sortChIC H3K27me3_ data from\n",
    "[Zeller, Yueng et. al. (2023)](https://www.nature.com/articles/s41588-022-01260-3). We aggregate the signal\n",
    "in 10kb bins.\n",
    "\n",
    "Our data is located in the ``sortchic_testdata`` directory. We can store the paths to our bamfiles and\n",
    "to the file containing our cell barcodes using the following commands. Note that the cell barcodes must\n",
    "be stored in a list of strings, where each string corresponds to one barcode.\n",
    "\n",
    "Additionally, we need a blacklist of problematic regions in the genome to filter out sequencing artifacts.\n",
    "We provide a blacklist for the mm10 mouse genome in our example data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b4d47ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = 'sortchic_testdata'\n",
    "bam_files = [bam_file.path for bam_file in os.scandir(data_dir) if bam_file.path.endswith('.bam')]\n",
    "\n",
    "barcodes_path = os.path.join(data_dir, 'sortChIC_barcodes.txt')\n",
    "barcodes = [line.strip() for line in open(barcodes_path)]\n",
    "\n",
    "blacklist = os.path.join(data_dir, 'mm10_blacklist.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e431693e",
   "metadata": {},
   "source": [
    "[CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin) takes the following arguments:\n",
    "- The list of paths to the bamfiles\n",
    "- Optionally, a region to limit the counting to in the format ``chrom:start:end``\n",
    "- The length of the bins to aggregate the signal into\n",
    "- The step size between bins. This is the distance between the start of two consecutive bins.\n",
    "Set equal to ``binLength`` to have non-overlapping bins covering the whole region.\n",
    "- The list of barcodes to count cells for, where each barcode corresponds to a cell in each sample.\n",
    "- The BAM tag where the cell barcodes are stored\n",
    "- The path to the blacklist BED/GTF file\n",
    "- The number of cores to use in the counting\n",
    "\n",
    "We instantiate [CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin) with\n",
    "all the necessary parameters, and perform the counting by calling ``CountReadsPerBin.run()``.\n",
    "This returns our _bin x cell_ count matrix and numpy array of bin IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c439b6f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "counter_bins = CountReadsPerBin(bam_files, region='chr1', binLength=50000, stepSize=50000,\n",
    "                                barcodes=barcodes, cellTag='BC', blackListFileName=blacklist,\n",
    "                                numberOfProcessors=16)\n",
    "bin_counts, bin_IDs = counter_bins.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d0876ba6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Counted 1536 barcodes and 3923 bins in chr1:47000000:47300000.\n",
      "Total counts: 14754957.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"\"\"\n",
    "Counted {bin_counts.shape[1]} barcodes and {bin_counts.shape[0]} bins in {region}.\n",
    "Total counts: {bin_counts.sum()}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e70ff916",
   "metadata": {},
   "source": [
    "Be mindful that each column output matrix contains the data for one barcode in each bamfile. The\n",
    "barcodes may not be present, or empty of counts, so filtering the count matrix is required to keep\n",
    "only barcodes corresponding to real cells. We provide a CLI tool for this purpose,\n",
    "[scCountQC](../tools/scCountQC.rst), and an example of its usage in\n",
    "[this tutorial](sincei_tutorial_sortChIC.rst).\n",
    "\n",
    "We can store our count matrix in an AnnData object for later processing and analysis.\n",
    "\n",
    "Note that ``AnnData`` stores data in _cell x feature_ format, so we must transpose the data matrix when assigning it to\n",
    "``adata.X``. Additionally, we can use ``bam_files``, ``barcodes``, and ``bin_IDs`` to add columns to the observation\n",
    "and variable dataframes of our ``AnnData.``"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8d7d9680",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Observation names and columns\n",
    "sample_names = [Path(bam_file).stem for bam_file in bam_files]\n",
    "samples = [sample for sample in sample_names for _ in range(len(barcodes))]\n",
    "cell_IDs = [f\"{sample}::{barcode}\" for sample in sample_names for barcode in barcodes]\n",
    "\n",
    "# Variable names and columns\n",
    "var_names = [var.split('::')[0] for var in bin_IDs]\n",
    "chroms = [var.split('_')[0] for var in var_names]\n",
    "starts = [int(var.split('_')[1]) for var in var_names]\n",
    "ends = [int(var.split('_')[2]) for var in var_names]\n",
    "\n",
    "# Create anndata object\n",
    "sortChIC_adata = ad.AnnData(\n",
    "    X=bin_counts.T,\n",
    "    obs={'sample': samples, 'barcode': barcodes * len(sample_names)},\n",
    "    var={'chrom': chroms, 'start': starts, 'end': ends},\n",
    "    )\n",
    "sortChIC_adata.obs_names = cell_IDs\n",
    "sortChIC_adata.var_names = bin_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c0e7f0ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACACTA</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACACTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACATAG</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACATAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACGAGA</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACGAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACTATC</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACTATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sortChIC-BM-SL2-k4me1-1::ACACTGAT</th>\n",
       "      <td>sortChIC-BM-SL2-k4me1-1</td>\n",
       "      <td>ACACTGAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    sample   barcode\n",
       "sortChIC-BM-SL2-k4me1-1::ACACACTA  sortChIC-BM-SL2-k4me1-1  ACACACTA\n",
       "sortChIC-BM-SL2-k4me1-1::ACACATAG  sortChIC-BM-SL2-k4me1-1  ACACATAG\n",
       "sortChIC-BM-SL2-k4me1-1::ACACGAGA  sortChIC-BM-SL2-k4me1-1  ACACGAGA\n",
       "sortChIC-BM-SL2-k4me1-1::ACACTATC  sortChIC-BM-SL2-k4me1-1  ACACTATC\n",
       "sortChIC-BM-SL2-k4me1-1::ACACTGAT  sortChIC-BM-SL2-k4me1-1  ACACTGAT"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sortChIC_adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3222d399",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_100760500_100810500::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100760500</td>\n",
       "      <td>100810500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_100810500_100860500::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100810500</td>\n",
       "      <td>100860500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_100860500_100910500::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100860500</td>\n",
       "      <td>100910500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_100910500_100960500::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100910500</td>\n",
       "      <td>100960500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_100960500_101010500::None</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100960500</td>\n",
       "      <td>101010500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               chrom      start        end\n",
       "chr1_100760500_100810500::None  chr1  100760500  100810500\n",
       "chr1_100810500_100860500::None  chr1  100810500  100860500\n",
       "chr1_100860500_100910500::None  chr1  100860500  100910500\n",
       "chr1_100910500_100960500::None  chr1  100910500  100960500\n",
       "chr1_100960500_101010500::None  chr1  100960500  101010500"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sortChIC_adata.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e30d682f",
   "metadata": {},
   "source": [
    "## 2. Counting reads in genomic features\n",
    "\n",
    "In addition to aggregating signal in bins, we can count the read counts for a known feature set. We\n",
    "provide a BED/GTF file containing our regions of interest to\n",
    "[CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin)\n",
    "instead of ``binLength`` and ``stepSize``.\n",
    "\n",
    "We will produce a count matrix of scATAC-seq signal in chromatin accessibility peaks for a region of\n",
    "chromosome 1, and a scRNA-seq count matrix for the genes in the same region. The data comes from a\n",
    "subset of cells of the 10x multiome dataset published in\n",
    "[Persad et. al.Â (2023)](https://www.nature.com/articles/s41587-023-01716-9)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "17a33e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '10x_multiome_testdata'\n",
    "\n",
    "barcodes_path = os.path.join(data_dir, '10x_barcodes_tutorial.txt')\n",
    "barcodes = [line.strip() for line in open(barcodes_path)]\n",
    "\n",
    "blacklist = os.path.join(data_dir, 'hg38_blacklist.v2.bed')\n",
    "\n",
    "region = 'chr1:47000000:47300000'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e869018",
   "metadata": {},
   "source": [
    "### scATAC-seq counts\n",
    "The file ``atacPeaks_replicateMerged.bed``contains chromatin accessibility peaks. Using it as input\n",
    "for [CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin) overrides the\n",
    "bin counting parameters. The rows in the resulting matrix each correspond to one region in the file.\n",
    "\n",
    "We specify that counting should be performed only in our regions of interest, so only peaks within\n",
    "the region will be included."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9ac54fa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_files = [bam_file.path for bam_file in os.scandir(data_dir)\n",
    "             if ('atac' in bam_file.path) and bam_file.path.endswith('.bam')]\n",
    "\n",
    "peaks_bed = os.path.join(data_dir, 'atacPeaks_replicateMerged.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4fb12c62",
   "metadata": {},
   "outputs": [],
   "source": [
    "counter_atac = CountReadsPerBin(\n",
    "    bam_files,\n",
    "    region = region,\n",
    "    bedFile = peaks_bed,\n",
    "    blackListFileName=blacklist,\n",
    "    barcodes = barcodes,\n",
    "    cellTag = 'CB',\n",
    "    numberOfProcessors=16\n",
    "    )\n",
    "atac_peak_counts, atac_peaks = counter_atac.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a0c624c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Counted 4256 barcodes and 45 peaks in chr1:47000000:47300000.\n",
      "Total counts: 32980.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"\"\"\n",
    "Counted {atac_peak_counts.shape[1]} barcodes and {atac_peak_counts.shape[0]} peaks in {region}.\n",
    "Total counts: {atac_peak_counts.sum()}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7b16ede",
   "metadata": {},
   "source": [
    "### scRNA-seq\n",
    "As with our with counting our scATAC-seq reads, we provide a file with regions for counting, in this\n",
    "case a GTF file containing all the genes in the HG38 human genome assembly. This file can be\n",
    "downloaded [here](http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.refGene.gtf.g).\n",
    "\n",
    "The file ``atacPeaks_replicateMerged.bed``contains chromatin accessibility peaks. Using it as input\n",
    "for [CountReadsPerBin](../modules/ReadCounter.rst#sincei.ReadCounter.CountReadsPerBin) overrides the\n",
    "bin counting parameters. The rows in the resulting matrix each correspond to one region in the file.\n",
    "\n",
    "We specify that counting should be performed only in our regions of interest, so only peaks within\n",
    "the region will be included."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "52809d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_files = [bam_file.path for bam_file in os.scandir(data_dir)\n",
    "             if ('gex' in bam_file.path) and bam_file.path.endswith('.bam')]\n",
    "\n",
    "genes_gtf = os.path.join(data_dir, 'hg38.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fe7155e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "counter_genes = CountReadsPerBin(\n",
    "    bam_files,\n",
    "    region = region,\n",
    "    bedFile = genes_gtf,\n",
    "    blackListFileName=blacklist,\n",
    "    barcodes = barcodes,\n",
    "    cellTag = 'CB',\n",
    "    numberOfProcessors=16\n",
    "    )\n",
    "rna_gene_counts, genes = counter_genes.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "327e09f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Counted 4256 barcodes and 19 genes in chr1:47000000:47300000.\n",
      "Total counts: 9890.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(f\"\"\"\n",
    "Counted {rna_gene_counts.shape[1]} barcodes and {rna_gene_counts.shape[0]} genes in {region}.\n",
    "Total counts: {rna_gene_counts.sum()}\n",
    "\"\"\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
